flybywiresim/a32nx
# feat(hyd): custom autobrake

Fixes #4290 
Fixes #4771 
Fixes #3042 
Fixes #2696
Fixes #5031
Fixes #5099

## Summary of Changes

Implements a new autobrake integrated in new hydraulic brakes system.

- [x] Replace current button logic
- [x] Integrate into hydraulic code properly
- [x] Deceleration governor
- [x] Arming / disarming logic including different pedal travels in different modes
- [x] Dynamic decel target vs time
- [x] Included key events LO MED and HI

## Screenshots (if necessary)



## References

![autobrake_fcom](https://user-images.githubusercontent.com/38904654/120675424-575fcf00-c495-11eb-96a1-7e621415f0e9.png)

`MED mode sends progressive pressure to the brakes  2 s after the  ground spoilers deploy in  order to decelerate the aircraft at  3 m/s² (9.8 ft/s²) 
LO mode  sends  progressive  pressure  to  the  brakes  2 s  after  the  ground  spoilers  deploy,  in order to decelerate the aircraft at  2 m/s² (6.6 ft/s²).`




## Additional context



Discord username (if different from GitHub):

## Testing instructions

New LVARS of custom autobrake are:
- A32NX_AUTOBRAKES_ARMED_MODE: 0 none /1 low / 2 med /3 max
- A32NX_AUTOBRAKES_DECEL_LIGHT:  AB is decelerating and has reached decel target (lights the blue DECEL light on the mode button)

1. Check autobrake can be armed on ground and stays armed when braking, unless A/skid is off or green pressure is lost.
2. Check autobrake decelerates if spoilers are armed and deploy on a rejected take off.
3. Check autobrakes low and medium decelerate 2 seconds after ground spoiler deployment on landing.
4. Check autobrakes deactivates from medium and low mode if decelerating and one pedal is above 53% or two pedals above 11%
5. Check autobrakes deactivates from max mode if decelerating and one pedal is above 77% or two pedals above 53%
6. Check autobrakes disarms after 10s in flight
7. Check max setting can't be armed after 10s in flight
8. Check that green decel light is on for the corresponding button mode when it is braking at max desired force. It might not light for med or max on slippery conditions (snow/ice).
9. Check that key or button assigned to LO /MED /HI autobrake command in input configuration works

Global verifications:
No strange brake behaviour (oscillations for example) on various weight configurations. Final decel tuning will be done after brake performance PR is pushed.
Brake temps are consistent with temperatures obtained with manual braking of same average force



## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
